File: src/main/java/io/legado/app/adapters/DefaultAdpater.kt
package io.legado.app.adapters

import io.legado.app.help.http.StrResponse
import io.legado.app.model.DebugLog
import com.htmake.reader.utils.getRelativePath
import java.nio.file.Paths

/**
 * Default implementation of ReaderAdapterInterface using existing getWorkDir functions.
 */
class DefaultAdpater : ReaderAdapterInterface {

    override fun getWorkDir(subPath: String): String {
        var workDirPath = ""
        val osName = System.getProperty("os.name")
        val currentDir = System.getProperty("user.dir")
        if (osName.startsWith("Mac OS", true) && !currentDir.startsWith("/Users/")) {
            workDirPath = Paths.get(System.getProperty("user.home"), ".reader").toString()
        } else {
            workDirPath = currentDir
        }
        return Paths.get(workDirPath, subPath).toString()
    }

    override fun getWorkDir(vararg subDirFiles: String): String {
        return getWorkDir(getRelativePath(*subDirFiles))
    }

    fun getRelativePath(vararg subDirFiles: String): String {
        return com.htmake.reader.utils.getRelativePath(*subDirFiles)
    }

    override fun getCacheDir(): String {
        return getWorkDir("storage", "cache")
    }

    override suspend fun getStrResponseByRemoteWebview(
        url: String?,
        html: String?,
        encode: String?,
        tag: String?,
        headerMap: Map<String, String>?,
        sourceRegex: String?,
        javaScript: String?,
        proxy: String?,
        post: Boolean,
        body: String?,
        userNameSpace: String,
        debugLog: DebugLog?
    ): StrResponse? {
        throw Exception("不支持webview")
    }
}

File: src/main/java/io/legado/app/adapters/DefaultAdpater.rs
// package io.legado.app.adapters
//
// import io.legado.app.help.http.StrResponse
// import io.legado.app.model.DebugLog
// import com.htmake.reader.utils.getRelativePath
// import java.nio.file.Paths

fn get_relative_path(sub_dir_files: &[&str]) -> String {
    let mut p = std::path::PathBuf::new();
    for f in sub_dir_files {
        p.push(f);
    }
    return p.to_string_lossy().to_string();
}

/**
 * Default implementation of ReaderAdapterInterface using existing getWorkDir functions.
 */
pub struct DefaultAdpater;

impl ReaderAdapterInterface for DefaultAdpater {

    fn get_work_dir(&self, sub_path: &str) -> String {
        let mut work_dir_path = String::new();
        let os_name = std::env::var("os.name").unwrap_or_default();
        let current_dir = std::env::var("user.dir").unwrap_or_default();
        if os_name.to_lowercase().starts_with("mac os") && !current_dir.starts_with("/Users/") {
            work_dir_path = std::path::Path::new(&std::env::var("user.home").unwrap_or_default())
                .join(".reader")
                .to_string_lossy()
                .to_string();
        } else {
            work_dir_path = current_dir;
        }
        return std::path::Path::new(&work_dir_path).join(sub_path).to_string_lossy().to_string();
    }

    fn get_work_dir_vararg(&self, sub_dir_files: &[&str]) -> String {
        return self.get_work_dir(&get_relative_path(sub_dir_files));
    }

    fn get_relative_path(&self, sub_dir_files: &[&str]) -> String {
        return get_relative_path(sub_dir_files);
    }

    fn get_cache_dir(&self) -> String {
        return self.get_work_dir_vararg(&["storage", "cache"]);
    }

    async fn get_str_response_by_remote_webview(
        &self,
        url: Option<&str>,
        html: Option<&str>,
        encode: Option<&str>,
        tag: Option<&str>,
        header_map: Option<&std::collections::HashMap<String, String>>,
        source_regex: Option<&str>,
        java_script: Option<&str>,
        proxy: Option<&str>,
        post: bool,
        body: Option<&str>,
        user_name_space: &str,
        debug_log: Option<&DebugLog>,
    ) -> Option<StrResponse> {
        panic!("不支持webview")
    }
}
